### DISCUSSION TITLE
Is the network graph applied to the localhost loopback connections?

### DISCUSSION THREAD

**Discussion by haxxpop** [shadow/shadow#2676] [Q&A]
I think the loopback connections should have no limit on bandwidth and nearly zero latency. Are the bandwidth limit, latency and the packet loss rate specified in the network graph applied to the loopback connection?

**Comment by stevenengler** [comment]
No they should not be applied to the loopback connection. There is a 1 ns latency for localhost packets, but hopefully we can remove that in the future.

See the `is_local` check and the surrounding code:

https://github.com/shadow/shadow/blob/89442dcd17b7a00f0d848c014001980495eabafa/src/main/host/network_interface.c#L484-L487

The packet loss occurs in `worker_sendPacket`, but this function is not called for localhost packets.

(We might be changing this code soon in #2632.)

**Reply by stevenengler** [reply]
It looks like there might be a bug in a specific case (a host sends packets to itself using its public IP rather than 127.0.0.1) that can cause the rate limit to be applied to localhost packets if the host is also sending a lot of non-localhost packets. #2677

**Reply by stevenengler** [reply]
Also, Shadow currently uses delayed ACKs on localhost, which can occasionally make localhost packets have a larger-than-expected 1 ms delay. #2683
